components/fraud_proof.md

```diff
@@ -0,0 +1,64 @@
+# Fraud Proof Algorithm for Simple Sequencer
+This is a WIP -- mostly just a few notes that I'm jotting down to not forget.
+
+## Requirements
+1. Not all blockhashes are posted up front.
+    - Blockhashes should be posted roughly every 100 L1 blocks.
+2. Only one block is evaluated during the fraud proof.
+
+## Fraud Proof Structure
+
+Fraud proofs are initialized with some constant size input. In our case this is:
+
+```python
+def init_fraud_proof(latest_blockhash: bytes, invalid_epoch_commitment: int):
+    ''' Invalid epoch commitment is the index of the invalid blockhash '''
+    # ...
+```
+
+What makes rollups special is that all of the preimages to the `latest_blockhash` are known. This means that from within the fraud proof VM, all blockhash preimages (including the state root, pre blockhash, etc) can be queried using the `preimage oracle`. Fundamentally this means that all historical blocks, events, transactions, and state can are accessible within the fraud proof VM.
+
+### Preimage Oracle
+
+The fraud proof VM has access to the following function (called the "preimage oracle"):
+
+```python
+def get_preimage(hash: bytes) -> bytes:
+    # The preimage is looked up from a local database held by the fraud prover
+    preimage = fetch_preimage(hash)
+    assert keccak256(preimage) == hash
+    return preimage
+```
+
+This seemingly simple functionality has extreme consequences. All publically available preimages can be operated on from within the fraud proof -- a pretty amazing property.
```

Comment:
Good note, but spec language should be more tempered. 

The pre-image oracle is neat and minimal, but also implies there is some database index that can resolve the `fetch_preimage` calls. Blockhashes/tx-hashes are easily found, other preimages may not be (e.g. historical intermediate internal MPT node hashes).

I like to think of the oracle as a substitute for the real DB, with two modes:
- during generation: fetch from the DB, and remember the execution step required the preimage (not striclty required to check if it matches, the DB is trusted, but a sanity-check is still nice to have)
- during verification: fetch from a list of pre-images specific to the step, and verify it matches.

The "fetch from DB" part is limited to interface of the DB we are using.

Reply:
Agree with everything you said! Maybe that would be a good framing to put into the writeup.

Also,
> spec language should be more tempered.
😄 Yes agreed. I think that as I was throwing this doc together I was just coming to realize how much we can do with the preimage oracle, things like searching back through chain history, and so seems my excitement bled into the doc 😂